ratchet-tree.go

```diff
@@ -0,0 +1,479 @@
+package mls
+
+import (
+	"bytes"
+	"fmt"
+	"github.com/bifurcation/mint/syntax"
+)
+
+//// Tree Hashes
+type ParentNodeInfo struct {
+	PublicKey      *HPKEPublicKey
+	UnmergedLeaves []leafIndex `tls:"head=4"`
+}
+
+type ParentNodeHashInput struct {
+	HashType  uint8
+	Info      *ParentNodeInfo `tls:"optional"`
+	LeftHash  []byte          `tls:"head=1"`
+	RightHash []byte          `tls:"head=1"`
+}
+
+type LeafNodeInfo struct {
+	PublicKey  HPKEPublicKey
+	Credential Credential
+}
+
+type LeafNodeHashInput struct {
+	HashType uint8
+	Info     *LeafNodeInfo `tls:"optional"`
+}
+
+//// ratchet tree node and helpers
+type OptionalRatchetNode struct {
+	Node *RatchetTreeNode
+	Hash []byte `tls:"head=2"`
+}
+
+func (n *OptionalRatchetNode) setLeafHash(cs CipherSuite) {
+	lhi := LeafNodeHashInput{HashType: 0}
+	if n.Node != nil {
+		p := n.Node.PublicKey
+		c := n.Node.Cred
+		if c == nil {
+			panic(fmt.Errorf("mls.rtn: Leaf node not provisioned with a credentialp"))
+		}
+		lhi.Info = &LeafNodeInfo{
+			PublicKey:  *p,
+			Credential: *c,
+		}
+	}
+
+	h, err := syntax.Marshal(lhi)
+	if err != nil {
+		panic(fmt.Errorf("mls.rtn: Marshal error %v", err))
+	}
+	n.Hash = cs.digest(h)
+
+}
+
+func (n *OptionalRatchetNode) merge(o *RatchetTreeNode) {
+	if n.Node == nil {
+		n.Node = o
+	} else {
+		n.Node.Merge(o)
+	}
+}
+
+func (n *OptionalRatchetNode) setHash(cs CipherSuite, l OptionalRatchetNode, r OptionalRatchetNode) {
+	phi := ParentNodeHashInput{HashType: 1}
+	if n.Node != nil {
+		phi.Info = &ParentNodeInfo{PublicKey: n.Node.PublicKey, UnmergedLeaves: n.Node.UnmergedLeaves}
+	}
+	phi.LeftHash = l.Hash
+	phi.RightHash = r.Hash
+	data, err := syntax.Marshal(phi)
+	if err != nil {
+		panic(fmt.Errorf("mls.rtn: set hash error %v", err))
+	}
+	n.Hash = cs.digest(data)
+}
+
+func (n *OptionalRatchetNode) hasPrivate() bool {
```

Comment:
This seems unnecessary.  The `checkInvariant` function can be deleted; it's only used in tests.